Refactor user and identity into AuthIdentityHandler instance variables

RyanSkonnord opened the pull request:
Broken out from https://github.com/getsentry/sentry/pull/30356. See [this comment](https://github.com/getsentry/sentry/pull/30356#issuecomment-985086072) for context.

Move the operation of looking up an existing user from the email address in the identity up front to the creation of the `AuthIdentityHandler` object. This prevents the need to pass around the "acting user" among methods, as newly required for the fix in #30356.

Business logic within `handle_unknown_identity` around "blanking" the acting user to None, thereby triggering a new account provisioning (sometimes), is now represented with a more meaningful boolean variable.

The signature change to `AuthIdentityHandler` requires changes in `getsentry`.

RyanSkonnord committed:
Refactor user and identity into AuthIdentityHandler instance variables

Move the operation of looking up an existing user from the email address
in the identity up front to the creation of the AuthIdentityHandler
object. This prevents the need to pass around the "acting user" among
methods later.

Business logic within handle_unknown_identity around "blanking" the
acting user to None, thereby triggering a new account provisioning
(sometimes), is now represented with a more meaningful boolean variable.

RyanSkonnord commented:
Similarly to [here](https://github.com/getsentry/sentry/pull/30356#issuecomment-987459513), I believe the existing test cases in `test_helper.py` suffice to cover these changes. Let me know if you disagree.

JoshFerge commented on src/sentry/auth/helper.py:
```diff
@@ -90,20 +90,21 @@ def _using_okta_migration_workaround(
+        self.user: User = (
+            self.request.user
+            if self.request.user.is_authenticated
+            else (self._get_user() or self.request.user)
```
maybe worth adding a comment, had to read it a few times before it made sense 😅  -- something like 
# prefer a logged in user, try to look up a user by email, or the anonymous user

RyanSkonnord replied:
👍 I originally had this broken out into separate if-else statements but wanted to combine it into one expression so that the type hint could be on a single assignment. But it occurred to me that, now that `_get_user` is called from only a single place, we could fold the `is_authenticated` check into it: https://github.com/getsentry/sentry/pull/30463/commits/bfa517f661684d72c9506e4eb68d18549eb30268

RyanSkonnord committed:
Refactor post_init logic into initialize_user

RyanSkonnord commented:
Test failures in `test_auth_organization_login` are interesting. The user is showing up as unauthenticated in the second login -- not sure whether that's a regression or a problem with the test. Will continue working on that; see also https://github.com/getsentry/sentry/pull/30356#issuecomment-987668163

RyanSkonnord committed:
Need _dispatch_to_confirmation to distinguish request user vs existing user

RyanSkonnord committed:
Debug login op

RyanSkonnord committed:
Fix method calls

RyanSkonnord commented:
Bugs should now be fixed. Not super happy with the haphazard way the code is mixing references to `self.user` and `self.request.user` in close proximity -- going to see if I can refactor for more clarity.

RyanSkonnord committed:
Re-simplify AuthIdentityHandler's user initialization

RyanSkonnord committed:
Introduce _logged_in_user property

Cuts down on the number of times we have to refer directly to
`self.request.user` and adds a little abstraction around
`is_authenticated` to make internal code cleaner.

JoshFerge reviewed (approved):
nice changes!
